### PR TITLE
ChangedStorage

### PR DESCRIPTION
**Pull request by Aceeri**
Extension on the discussions in gitter and https://github.com/slide-rs/specs/issues/89.

### CODE REVIEW

File: src/storage/tests.rs [hunk 1]
```diff
@@ -451,19 +467,34 @@ mod test {
...
         w.register_with_id::<FlaggedCvec>(1);
         w.register_with_id::<FlaggedCvec>(2);
-        let mut s1: Storage<FlaggedCvec, _> = w.write_with_id(1);
-        let mut s2: Storage<FlaggedCvec, _> = w.write_with_id(2);
 
-        for i in 0..15 {
-            // Test insertion flagging
-            s1.insert(Entity::new(i, Generation::new(1)), i.into());
-            assert!(s1.open().1.flagged(Entity::new(i, Generation::new(1))));
+        let mut entities = Vec::new();
```
Comment (torkleyy) [inline src/storage/tests.rs]: Why are you doing this complicated loop here? Wouldn't a simple `.collect()` be sufficient? Or given that you only use these entities once, no allocation at all?

File: src/storage/storages.rs [hunk 2]
```diff
@@ -332,3 +334,191 @@ impl<'a, C, T: UnprotectedStorage<C>> Join for &'a mut FlaggedStorage<C, T> {
...
+        }
+    }
+    // Forwarding methods to `FlaggedStorage`.
+    /// Whether the component related to the entity was flagged or not.
+    pub fn flagged<E: EntityIndex>(&self, entity: E) -> bool {
+        self.inner.flagged(entity)
+    }
+    /// Clears the bitset for flagged components.
+    ///
+    /// Should be called at least once depending on where you want to reset the flags.
```
Comment (torkleyy) [inline src/storage/storages.rs]: I think it should be always called at the beginning of the system which actually sets these components.

File: src/storage/storages.rs [hunk 3]
```diff
@@ -332,3 +334,191 @@ impl<'a, C, T: UnprotectedStorage<C>> Join for &'a mut FlaggedStorage<C, T> {
...
+          U: UnprotectedStorage<C>,
+{
+    /// Maintains the flagging and cache of the `ChangedStorage`.
+    /// 
+    /// Checks for equality between the cache and the actual components data
+    /// to filter out the components that did not change, leaving only the modified ones
+    /// in the flagged bitset.
+    ///
+    /// Should be called whenever you care about flagged components or alternatively
+    /// whenever the component storage is modified.
```
Comment (torkleyy) [inline src/storage/storages.rs]: This is very unclear; unless you know the implementation details, this line won't help users. Could you maybe include a usage example?

**Comment by torkleyy** [comment]
Given that this is a rather big feature and needs to be carefully reviewed, we decided to wait for this until `0.9` is published.

File: src/storage/storages.rs (hunk 3 above)
Comment (Aceeri) [inline src/storage/storages.rs]: This method doesn't matter anymore.

File: src/storage/storages.rs (hunk 2 above)
Comment (Aceeri) [inline src/storage/storages.rs]: There are no right times to clear the flags, it depends on the context. If every system that sets these components to something clears them then the flags are completely useless to other systems for things like transforms. In other circumstances if you only want it to be flagged from a certain iteration then you can clear it before that iteration.

**Comment by torkleyy** [comment]
@Aceeri I think you implemented a `TrackedStorage` in the past. Do you think we'd need both (`ChangedStorage` and `TrackedStorage`)?

**Comment by Aceeri** [comment]
@torkleyy The `TrackedStorage` stuff I implemented was just `FlaggedStorage` with insertion/removal of components tracking. The `ChangedStorage` is still kind of a more "correct" flagged storage since it only flags when something was truly modified, but I'm unsure if that level of accuracy is worth the complexity still (considering the whole issue I had with component wrapper storages vs addon storage).

**Comment by Aceeri** [comment]
This storage seems kind of niche, so I'm probably not going to be working on it for a while. If anyone else wants to pick this up and use it themselves it should work.

**Pull request closed without merging**